client/state/data-layer/wpcom/sites/rewind/index.js

```diff
@@ -27,17 +33,75 @@ const fetchRewindState = action =>
 		action
 	);
 
+/**
+ * Re-requests the activity log entries for all
+ * previously-sent queries in application state
+ *
+ * This function is highly coupled to the details
+ * of Activity Log and Rewind state!
+ *
+ * It will not grab events we haven't already
+ * gotten. Instead it will reissue the same reqeusts
+ * as previously made so that we get updates on
+ * things like the `isDiscarded` status of events.
+ *
+ * @param {Number} siteId the site being queried
+ * @param {Object} rewind Rewind state for the site
+ * @returns {function} thunk action maybe dispatching requests
+ */
+const refreshActivityLogAfterRewind = ( siteId, rewind ) => ( dispatch, getState ) => {
+	if ( ! rewind || rewind.status === 'running' ) {
+		return;
+	}
+
+	const state = getState();
+	const prev = state.rewind[ siteId ];
+
+	if ( ! prev || ! ( prev.rewind && prev.rewind.status === 'running' ) ) {
+		return;
+	}
+
+	const logItems = state.activityLog.logItems[ siteId ];
+	if ( ! logItems ) {
+		return;
+	}
+
+	Object.keys( logItems.data.queries ).forEach( query =>
+		dispatch( activityLogRequest( siteId, fromPairs( JSON.parse( query ) ) ) )
+	);
+};
+
 const updateRewindState = ( { siteId }, data ) => {
 	const stateUpdate = {
 		type: REWIND_STATE_UPDATE,
 		siteId,
 		data,
 	};
 
-	const hasRunningRewind = data.rewind && data.rewind.status === 'running';
+	const hasRunningRewind =
+		data.rewind && ( data.rewind.status === 'queued' || data.rewind.status === 'running' );
 
 	if ( ! hasRunningRewind ) {
-		return stateUpdate;
+		return [
+			/*
+			 * When we finish a Rewind operation we may have
+			 * changed the `isDiscarded` value for previous
+			 * events in the Activity Log. Thus, we'll need to
+			 * reissue the requests which pulld all of the
+			 * previous events into Calypso.
+			 *
+			 * This _must_ go before the state update because
+			 * it relies on existing state to know if we have
+			 * transitioned from a time when we had a Rewind
+			 * operation running to a time when we no longer
+			 * have that.
+			 *
+			 * If we wait until after state is updatd then it
+			 * will never look like we had on.
+			 */
+			refreshActivityLogAfterRewind( siteId, data.rewind ),
```

Comment:
this is pretty cool. a thunk inside of the array of actions that get dispatched.

Reply:
in this case the important bit was accessing app state at the time it was called, something not available to the middleware otherwise.

the alternative is handling this in the UI which is also viable.

on one hand the data is related and invalidated by the finished Rewind operation so that in itself is enough to warrant purging the stale records. this is why we do it here.

on the other hand there's no need to load all that data until it's actually viewed. the data layer pre-empts network fetches needlessly.

downsides to doing it in the UI:
 - we may fail to detect the transition from a running operation to a non-running operation and thus fail to purge the stale records
 - we introduce visual lag in the redisplay. by performing it in the data layer we minimize the time that the stale information is displayed. waiting for the view means we more likely see a flash of events.

we could make this simpler and just clear out those records and let the UI reload them. I felt like the flash of no content would be jarring especially since we only expect a relatively small subset of visual elements to change (only a disabled-state on a subset of the events)

still we could have maintained a `lastKnownRewindStatus` here which I was starting to do at first but then thought it might be better to base the logic off of app-state. I'm not entirely sure that was the right decision and I don't see much effective difference between prepping app-state for a test and prepping this independent handler. actually I think the simpler `lastKnown…` approach might be better overall 🤷‍♂️